Topic 4: k Nearest Neighbours

## Task 4

Using the famous iris data set suggest whether the setosa class is easily separable from the other two classes. Provide evidence for your answer.

### Content

In this exercise we aim to train a machine learning model that can take in feature data from the iris dataset and test the accuracy of the model when it tries to predict which of the three flower varieties a set of features are from.

The inputs are recorded measurments of the flowers featurse under the column headings: sepal_length, sepal_width, petal_length and petal_width. The varieties are setosa, versicolor and verginica under the heading class.

###  scikit-learn

We will use the sikit-learn. Scikit-learn is a machine learning library in Python. It is built on NumPy and SciPy and it contains a wide range of tools for machine learning, including: classification, regression, clustering, and dimension reduction tasks. It is a popular choice for machine learning beginners as it has good documentation and a broad library of functions. It is an excellent platform for learning. The package in python is actually called `sklearn`. This shorter name is helpful as it avoids confusion with extra charachters. It is generally included with annaconda so no need to install it seperatley.

>Scikit-learn is an open source machine learning library that supports supervised and unsupervised learning. It also provides various tools for model fitting, data preprocessing, model selection, model evaluation, and many other utilities.

 [[401]](#ref401) (scikit-learn.org, Getting Started, Nov 2023)


### sikit-learn fit method

The scikit-learn `fit()` method is used to train or teach the learning model on a specific dataset. The model begins with default settings that may not be good at predicting the type of flower without any knowledge of flowers. Instead of using its previous knowlodge it learns the patterns in the data and uses those patterns to make predictions on new data. A training step with a fit method is required for most machine learning models before they can be used to make predictions.

## Start coding

In [27]:
#imports required to run the notebook code
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
import math

Lets have a look at the data and check for null values. 

In [28]:
iris = pd.read_csv(r"..\data\iris.csv")
iris.head()

sepal_length  sepal_width  petal_length  petal_width   class
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [29]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   class         150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [30]:
iris.isnull().values.any()

False

The data is quite short with just 150 entries but no null (or `NaN`) values. This is good, because we don't need to worry about our results being skewed by empty cells. We can see column headings from the data `head()` output. Varieties are under the heading `class`. The other feature headings can also be seen. We will need these too.

To carryout a split train and learn on the data we need separate out the target variable. In this case we are looking to predict class, so we seperate it out as y. The feature data we are using to predict the class is seperated to the variable X. After that we will split the dataset out into training (70%) and testing (30%) sets using the `train_test_split()` function from sikit-learn

In [31]:
# Separate out the target variable. In this case we are looking to predict class, so we seperate it out as y. 
# The feature data we are using to predict the class is seperated to the variable X. 
X = iris[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]
y = iris['class']

In [32]:
#https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
#sklearn.model_selection.train_test_split(*arrays, test_size=None, train_size=None, random_state=None, shuffle=True, stratify=None)

#Slpit the dataset into training (70%) and testing (30%) sets.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, shuffle=True)

Below using the data `head()` function again we can see the `X_train` and `X_test` sets contain the same data for different selected rows. Also notice the class that we wish to predice is now missing. This is seperated into the `y` datasets. 

In [33]:
X_train.head()

sepal_length  sepal_width  petal_length  petal_width
34            4.9          3.1           1.5          0.1
134           6.1          2.6           5.6          1.4
146           6.3          2.5           5.0          1.9
76            6.8          2.8           4.8          1.4
43            5.0          3.5           1.6          0.6

In [34]:
X_test.head()

sepal_length  sepal_width  petal_length  petal_width
62            6.0          2.2           4.0          1.0
29            4.7          3.2           1.6          0.2
125           7.2          3.2           6.0          1.8
139           6.9          3.1           5.4          2.1
143           6.8          3.2           5.9          2.3

K-Nearest Neighbors (KNN) is an algorithm that can classify new data points based on the majority class of its k nearest neighbors in the training set. 

The k-value is an important parameter in KNN. It determines the number of nearest neighbors considered. The best k-value will differ based on the dataset can be optimised by experimentation

[[402]](#ref402) (ibm.com, What is the k-nearest neighbors algorithm? Nov 2023)

>What is a good value for K in KNN?
How to find the optimal value of K in KNN? | by Amey Band ...
The optimal K value usually found is the square root of N, where N is the total number of samples. Use an error plot or accuracy plot to find the most favorable K value. KNN performs well with multi-label classes, but you must be aware of the outliers.23 May 2020

How to find the optimal value of K in KNN? 

[[403]](#ref403) (towardsdatasience.com, How to find the optimal value of K in KNN? Nov 2023)



In [35]:
knn_value = round(math.sqrt((len(iris.index))))

In [36]:
# Create and fit a KNeighborsClassifier model
knn = KNeighborsClassifier(n_neighbors=knn_value)
knn.fit(X_train, y_train)
print(knn.fit(X_train, y_train))

KNeighborsClassifier(n_neighbors=12)


In [37]:
accuracy = (knn.score(X_test, y_test))*100
accuracy

97.77777777777777

### Results

When the code evaluates on the testing set, it is does so with a mixture of all three classes. The code then uses its knowledge of each class to predict a class. As a result, the accuracy of the code is a measure of its ability to correctly classify all three varieties of flowers, not just Iris setosa.

[[404]](#ref404) (towardsdatasience.com, Importance of Distance Metrics in Machine Learning Modelling, Nov 2023)

In this case the model can predict a flowers variety or `class` with 97.77% accuracy.


### Results for setosa

Lets look at the results more closely to see how accuratley setosa can be predicted. For this we can use the sikitlearn `classification_report()` function. It will output a report of our models performance by giving scores for `precision`, `recall`, `F1-score`, and `support` for each class.

    Precision is the proportion of positive predictions that are correct

    Recall is the proportion of actual positives that are correctly identified

    F1-score is a measure of the balance between precision and recall

    Support is the number of actual occurrences of each class in the dataset

[[405]](#ref405) (kaggle.com/, kNN Classifier Tutorial, Nov 2023)

In [38]:
y_pred = knn.predict(X_test)

In [39]:
#https://medium.com/@mehtashubh1029/iris-flower-classification-using-knn-1eef6e7f3f84
#The classification_report function builds a text report showing the main classification metrics. 
#https://scikit-learn.org/stable/modules/model_evaluation.html#classification-report
print(classification_report(y_pred, y_test))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        10
  versicolor       1.00      0.93      0.97        15
   virginica       0.95      1.00      0.98        20

    accuracy                           0.98        45
   macro avg       0.98      0.98      0.98        45
weighted avg       0.98      0.98      0.98        45



### Conclusions

Using sikit-learn on the iris dataset we predicted a flowers variety or `class` with 97.77% accuracy. More importantly for this exercise we predicted `setosa` with 100% accuracy.

### References
[401]<a id='ref401'></a>  (scikit-learn.org, Getting Started, Nov 2023) https://scikit-learn.org/stable/getting_started.html

[402]<a id='ref402'></a>  (ibm.com, What is the k-nearest neighbors algorithm? Nov 2023) retreived from: https://www.ibm.com/topics/knn

[403]<a id='ref403'></a>  (towardsdatasience.com, How to find the optimal value of K in KNN? Nov 2023) retreived from: https://towardsdatascience.com/how-to-find-the-optimal-value-of-k-in-knn-35d936e554eb

[404]<a id='ref404'></a>  (towardsdatasience.com, Importance of Distance Metrics in Machine Learning Modelling, Nov 2023) retreived from:  https://towardsdatascience.com/importance-of-distance-metrics-in-machine-learning-modelling-e51395ffe60d

[405]<a id='ref405'></a>  (kaggle.com/, kNN Classifier Tutorial, Nov 2023) retreived from: https://www.kaggle.com/code/prashant111/knn-classifier-tutorial

### END task4

[Back to top of notebook](#Task-4)